### ISSUE TITLE
Bad location data in location and network probably due to losing GPS

### ISSUE THREAD

**Issue by mulaz** [wiglenet/wigle-wifi-wardriving#82]
I have a couple of issues with Wigle on a LG Nexus 4 phone (stock android 5.1.1). I'm not sure if it's device specific or a general issue. 

First and a minor one is that sometimes the "location" table contains wrong location data, eg:

```
sqlite> select _id,level,lat,lon,altitude,accuracy,time from location order by lat desc limit 100;
_id         level       lat         lon         altitude    accuracy        time      
----------  ----------  ----------  ----------  ----------  --------------  ----------
748         -87         Inf         -Inf        0.0         566.0791015625  0         
749         -91         Inf         -Inf        0.0         566.0791015625  0     
750         -86         Inf         -Inf        0.0         566.0791015625  0         
751         -90         Inf         -Inf        0.0         566.0791015625  0         
752         -91         Inf         -Inf        0.0         566.0791015625  0         
4252        -66         Inf         -Inf        0.0         0.924082756042  0         
...
```
"Inf" and "-Inf" are not valid WGS84 coordinates, also time=0 even if it's a valid number isn't OK/real. 
This issue is probably caused by saving data before obtaining a GPS fix or after losing it (not sure). 

There are also lines with valid coordinates (although i think the location is wrong - see below) and time=0.

```
_id         level       lat          lon               altitude    accuracy          time      
----------  ----------  -----------  ----------------  ----------  ----------------  ----------
12103       -85         46.06612801  14.5284741499999  0.0         9.52606964111328  0         
12104       -90         46.06603457  14.52453655       0.0         9.52606964111328  0         
```

This causes issues when trying to draw all the points on a map, but can be avoided by filtering out the invalid location when exporting data from sqlite. 

The bigger issue for me is the "network" table, where last- and best- locations are stored. 

I have visualised a part of best-location data (one line in 'network' table is one dot drawn with "bestlon", "bestlat" coordinates - http://i.imgur.com/DF3hHTc.png ). I have actually walked/scanned the bottom part, but the slightly right leaning horizontal "line of dots" is due to losing GPS/location errors described above. I never actually walked there. 

I have picked a known network/AP (larger, red dot, marked 'BSSID'), and visualised the data from 'location' table (each dot is a line from location table drawn at lat/lon coordinates; my selected AP is marked with larger red dots and 'BSSID - image: http://i.imgur.com/MtlGiGZ.png ). The actual location of the access point is in the bottom part of the image (close to where three more sighthings are drawn (larger red dots). The approximate distance between the three dots in the bottom part of the image, and the upper dot is ~1km.  Data: 

```
sqlite> select _id,level,lat,lon,altitude,accuracy,time from location where bssid="<BSSID>";
_id         level       lat          lon          altitude          accuracy    time         
----------  ----------  -----------  -----------  ----------------  ----------  -------------
24201       -81         46.04416077  14.48949131  348.299987792969  10.0        1482300334000
26714       -60         46.05357032  14.49049531  0.0               60.9652557  0            
26884       -69         46.0441834   14.48921632  337.899993896484  10.0        1482346381000
100093      -80         46.04402456  14.48904799  343.0             4.0         1483432558000

```

Three sightings have correct coordinates and valid GPS data (including time), but one, probably without a GPS fix (time = 0, and a wrong location - upper dot in the image) is also used for 'bestlat'/'bestlon' in 'network' table, since the signal level was the highest there. 

There are also lines in 'network' table where bestlat/bestlon or lastlat/lastlon are +/-Inf:

```
sqlite> select lasttime,lastlat,lastlon,bestlat,bestlat from network order by lastlat limit 5;
lasttime    lastlat     lastlon     bestlat     bestlat   
----------  ----------  ----------  ----------  ----------
0           -Inf        -Inf        -Inf        -Inf      
0           -Inf        -Inf        -Inf        -Inf      
0           -Inf        -Inf        -Inf        -Inf      
0           -Inf        -Inf        46.4612570  46.4612570
0           -Inf        Inf         46.0650862  46.0650862
```

```
sqlite> select lasttime,lastlat,lastlon,bestlat,bestlat from network order by bestlat limit 5;
lasttime    lastlat     lastlon     bestlat     bestlat   
----------  ----------  ----------  ----------  ----------
0           -Inf        -Inf        -Inf        -Inf      
0           -Inf        -Inf        -Inf        -Inf      
0           -Inf        -Inf        -Inf        -Inf      
1483223038  46.4615528  13.9437285  -Inf        -Inf      
1483112210  46.0550346  14.5191821  -Inf        -Inf      
```

So, sometimes, (eg.) when I walk into a building, and lose the GPS fix, it makes the location drift in a random direction and then store the invalid location to the database. Sometimes it even stores +/-Inf. If the access points signal is stronger there (and it often is inside buildings), the invalid location is also used for "bestlat" and "bestlon" fields.  I can also fix this in postprocessing (eg. foreach bssid in 'network' table, select all the sightings from the 'location' table, and pick the highest signal one with valid time and coordinates).

I was wondering if I was the only one with this issues (some weird hw/sw combination), or has anyone else noticed similar errors in data (location points going in random directions, and/or appearing in locations never visited, or +/-Inf coordinates in the 'network' table). Maybe wigle shouldn't store location (to any table), if any of the location data is invalid (eg. +/Inf in coordinates, time==0). I can fix my data when processing, but this issue probably also affects kml export, etc. for other people.

**Comment by bobzilladev** [comment]
We filter for this kind of thing on the server side on file uploads, but the in-app doesn't have much logic around that. We'd love pull requests taking on that problem, it probably wouldn't be a lot of lines but require a decent amount of testing since it's all heuristics. But just checking if signal or time are sane values would be pretty simple.

This is where the decision is made:
https://github.com/wiglenet/wigle-wifi-wardriving/blob/master/wiglewifiwardriving/src/main/java/net/wigle/wigleandroid/DatabaseHelper.java#L747

**Comment by rksh** [comment]
making the most basic change for now, avoiding updating network table with bad data, still recording locations. I'll double-check server-side logic next.

**Comment by bobzilladev** [comment]
This change is now released in 2.18, closing.
